controller/token.go

```diff
@@ -457,6 +459,56 @@ func (c *TokenController) Exchange(ctx *app.ExchangeTokenContext) error {
 			TokenType:    &keycloakToken.TokenType,
 		}
 
+		return ctx.OK(token)
+	} else if payload.GrantType == refreshToken {
```

Comment:
to avoid making this function grow huge, put it a new method and call it here.

Reply:
I can break Exchange method and put helping methods in login/service.go

Reply:
You don't have to put it to login/service.go. But the code will be more readable if you break Exchange method to three sub-methods. One for each grant type.
Moving code to a service makes sense if/when it helps you to test the code. It may allow you to mock services in the tests.

Reply:
Another reason for moving some code to a service is to make the code easier to reuse.
So, in this case it's probably makes sense to do both: split Exchange and have one method per grant type + move the part of refresh functionality which talks to remote KC to the login service. So you can mock this service in the tests and reuse it in the old refresh controller.

Reply:
https://github.com/fabric8-services/fabric8-auth/pull/272/commits/0fe53c5fc479aa8013bec637e0d8b46ac2cc558a#diff-6d42942038923aa03684ba83f0ab86fdR368
Exchange is now broken into three functions based on grant_type